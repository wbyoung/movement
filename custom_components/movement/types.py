"""Movement dataclasses and typing."""

from __future__ import annotations

from collections.abc import Generator
from dataclasses import dataclass, fields
import datetime as dt
from enum import Flag, StrEnum, auto
from typing import Any, Self, cast

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import State
from homeassistant.exceptions import IntegrationError
from homeassistant.util import dt as dt_util

from . import coordinator as cdn, statistics as sts

type MovementConfigEntry = ConfigEntry[cdn.MovementUpdateCoordinator]


class MisconfigurationError(IntegrationError):
    """Error to indicate that the integration is misconfigured."""


class EntityMissingError(IntegrationError):
    """Error to indicate that a tracked/dependent entity is missing."""

    def __init__(self, entity_id: str) -> None:
        """Initialize."""

        super().__init__(f"Entity could not be found for {entity_id}")


@dataclass
class SkipUpdateCondition(Exception):
    """Error condition to indicate that an update should be skipped."""

    reason: str


@dataclass
class TransitionRequiredCondition(Exception):
    """Error condition to indicate that a transition should begin."""

    reason: str


@dataclass
class MovementConfigEntryRuntimeData:
    """Runtime data definition."""

    coordinator: cdn.MovementUpdateCoordinator


class AbsentFalse(Flag):
    """Absent class with singleton for `False` value."""

    _singleton = False


class AbsentNone(Flag):
    """Absent class with singleton for `None` value."""

    _singleton = None


ABSENT_FALSE = AbsentFalse._singleton  # noqa: SLF001
ABSENT_NONE = AbsentNone._singleton  # noqa: SLF001


class ModeOfTransit(StrEnum):
    """ModeOfTransit class."""

    WALKING = auto()
    BIKING = auto()
    DRIVING = auto()


@dataclass(frozen=True, kw_only=True)
class StatisticGroup:
    """StatisticGroup class."""

    update_rate: sts.Statistic
    speed_recent_avg: sts.Statistic
    speed_recent_max: sts.Statistic

    def __iter__(self) -> Generator[sts.Statistic]:
        for field in fields(self):
            yield getattr(self, field.name)

    def items(self) -> Generator[tuple[str, sts.Statistic]]:
        for field in fields(self):
            yield (field.name, getattr(self, field.name))


class RecalculationRequest:
    """RecalculationRequest base class."""


@dataclass
class StateChangedData(RecalculationRequest):
    """StateChangedData class."""

    old_state: State
    new_state: State


@dataclass
class ServiceAdjustment(RecalculationRequest):
    """ServiceAdjustment class."""

    distance: float = 0
    adjustments: float = 0
    mode_of_transit: ModeOfTransit | None = None


@dataclass
class ResetRequest(RecalculationRequest):
    """ResetRequest class."""


@dataclass
class SpeedStaleIndicator(RecalculationRequest):
    """SpeedStaleIndicator class."""


@dataclass
class UpdatesStalledIndicator(RecalculationRequest):
    """UpdatesStalledIndicator class."""


@dataclass
class Location:
    """Location class for GPS location."""

    latitude: float
    longitude: float

    def as_dict(self) -> dict[str, Any]:
        return {
            "latitude": self.latitude,
            "longitude": self.longitude,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Self:
        return cls(
            latitude=data["latitude"],
            longitude=data["longitude"],
        )


@dataclass
class HistoryEntry:
    """HistoryEntry class."""

    NO_ACCURACY = float("inf")  # noqa: RUF045

    at: dt.datetime
    location: Location | AbsentNone
    accuracy: float
    ignore: str | AbsentNone = ABSENT_NONE
    debounce: bool | AbsentFalse = ABSENT_FALSE

    def as_dict(self) -> dict[str, Any]:
        result = {
            "at": self.at.isoformat(sep=" "),
            "accuracy": self.accuracy,
        }

        if self.location != ABSENT_NONE:
            result["location"] = self.location.as_dict() if self.location else None

        if self.ignore != ABSENT_NONE:
            result["ignore"] = self.ignore or None

        if self.debounce != ABSENT_FALSE:
            result["debounce"] = self.debounce or None

        return result

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Self:
        location = data.get("location", ABSENT_NONE)

        if location:
            location = Location.from_dict(location)

        return cls(
            at=dt_util.parse_datetime(data["at"]),
            location=location,
            accuracy=data.get("accuracy", HistoryEntry.NO_ACCURACY),
            ignore=data.get("ignore", ABSENT_NONE),
            debounce=data.get("debounce", ABSENT_FALSE),
        )


@dataclass
class TransitionEntry:
    """TransitionEntry class."""

    distance: float
    adjustments: float | AbsentNone = ABSENT_NONE

    def as_dict(self) -> dict[str, Any]:
        result = {
            "distance": self.distance,
        }

        if self.adjustments != ABSENT_NONE:
            result["adjustments"] = self.adjustments

        return result

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Self:
        return cls(
            distance=data["distance"],
            adjustments=data.get("adjustments", ABSENT_NONE),
        )


@dataclass
class TypedMovementData:
    """TypedMovementData class."""

    distance: float
    trip_distance: float | None
    trip_adjustments: float | None
    trip_start: dt.datetime | None = None

    def as_dict(self) -> dict[str, Any]:
        return {
            "distance": self.distance,
            "trip_distance": self.trip_distance,
            "trip_adjustments": self.trip_adjustments,
            "trip_start": self.trip_start.isoformat(sep=" ")
            if self.trip_start
            else None,
        }

    def as_state_dict(self) -> dict[str, Any]:
        return {
            "state": str(self.distance),
            "attributes": {
                "trip_distance": self.trip_distance,
                "trip_adjustments": self.trip_adjustments,
                "trip_start": self.trip_start.isoformat(sep=" ")
                if self.trip_start
                else None,
            },
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Self:
        return cls(
            distance=data["distance"],
            trip_distance=data.get("trip_distance"),
            trip_adjustments=data.get("trip_adjustments"),
            trip_start=dt_util.parse_datetime(data["trip_start"])
            if data.get("trip_start")
            else None,
        )


@dataclass
class MovementData:
    """MovementData class."""

    distance: float
    adjustments: float
    speed: float | None
    mode_of_transit: ModeOfTransit | None
    change_count: int  # changes to distance
    ignore_count: int  # ignored updates due to accuracy/debounce

    def as_dict(self) -> dict[str, Any]:
        return {
            "distance": self.distance,
            "adjustments": self.adjustments,
            "speed": self.speed,
            "mode_of_transit": self.mode_of_transit,
            "change_count": self.change_count,
            "ignore_count": self.ignore_count,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Self:
        return cls(
            distance=data.get("distance", 0),
            adjustments=data.get("adjustments", 0),
            speed=data.get("speed"),
            mode_of_transit=cast("ModeOfTransit", data.get("mode_of_transit")),
            change_count=data.get("change_count", 0),
            ignore_count=data.get("ignore_count", 0),
        )
