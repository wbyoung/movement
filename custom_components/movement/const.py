"""Constants for the Movement integration."""

import datetime as dt
from typing import Final

DOMAIN: Final = "movement"

ATTR_DISTANCE: Final = "distance_traveled"
ATTR_DISTANCE_ADJUSTMENTS: Final = "distance_adjustments"
ATTR_DISTANCE_WALKING: Final = "distance_walking"
ATTR_DISTANCE_BIKING: Final = "distance_biking"
ATTR_DISTANCE_DRIVING: Final = "distance_driving"
ATTR_SPEED: Final = "speed"
ATTR_MODE_OF_TRANSIT: Final = "mode_of_transit"
ATTR_GPS_ACCURACY: Final = "gps_accuracy"
ATTR_UPDATE_COUNT: Final = "distance_updates"

CONF_DEPENDENT_ENTITIES: Final = "dependent_entities"
CONF_TRACKED_ENTITY: Final = "tracked_entity"
CONF_TRIP_ADDITION: Final = "trip_addition"
CONF_MULTIPLIERS: Final = "multipliers"
CONF_NEIGHBORHOOD: Final = "neighborhood"
CONF_LOCAL: Final = "local"
CONF_HIGHWAY: Final = "highway"

UPDATES_STALLED_DELTA: Final = dt.timedelta(minutes=20)
HISTORY_EXPIRATION_DELTA: Final = dt.timedelta(hours=1)
SPEED_USABLE_DELTA: Final = dt.timedelta(seconds=45)
DEBOUNCE_UPDATES_DELTA: Final = dt.timedelta(seconds=5)

MAX_RESTORE_HISTORY: Final = 25
MAX_RESTORE_TRANSITION: Final = 25
