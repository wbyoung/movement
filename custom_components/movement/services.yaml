add_distance:
  fields:
    config_entry:
      required: true
      selector:
        config_entry:
          integration: movement
    distance:
      required: true
      selector:
        number:
          min: -40075
          max: 40075
          mode: box
          unit_of_measurement: "km"
    adjustments:
      required: false
      selector:
        number:
          min: -40075
          max: 40075
          mode: box
          unit_of_measurement: "km"
    mode_of_transit:
      required: false
      selector:
        text:
