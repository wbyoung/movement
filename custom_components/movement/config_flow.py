"""Config flow for Movement integration."""

from __future__ import annotations

from typing import Any, Final, cast

from homeassistant.components.device_tracker import DOMAIN as DEVICE_TRACKER_DOMAIN
from homeassistant.components.person import DOMAIN as PERSON_DOMAIN
from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN, SensorStateClass
from homeassistant.components.template import DOMAIN as TEMPLATE_DOMAIN
from homeassistant.config_entries import (
    ConfigEntry,
    ConfigFlow,
    ConfigFlowResult,
    OptionsFlow,
)
from homeassistant.const import UnitOfLength
from homeassistant.core import State, callback
from homeassistant.data_entry_flow import section
from homeassistant.helpers import entity_registry as er
from homeassistant.helpers.selector import (
    EntitySelector,
    EntitySelectorConfig,
    NumberSelector,
    NumberSelectorConfig,
    NumberSelectorMode,
)
from homeassistant.helpers.typing import VolDictType
from homeassistant.util import slugify
import voluptuous as vol

from .const import (
    CONF_DEPENDENT_ENTITIES,
    CONF_HIGHWAY,
    CONF_LOCAL,
    CONF_MULTIPLIERS,
    CONF_NEIGHBORHOOD,
    CONF_TRACKED_ENTITY,
    CONF_TRIP_ADDITION,
    DOMAIN,
)

SECTION_ADVANCED_OPTIONS: Final = "advanced_options"


def _base_schema(user_input: dict[str, Any]) -> VolDictType:
    return {
        vol.Required(
            CONF_TRACKED_ENTITY, default=user_input.get(CONF_TRACKED_ENTITY, [])
        ): EntitySelector(
            EntitySelectorConfig(domain=[DEVICE_TRACKER_DOMAIN, PERSON_DOMAIN]),
        ),
        vol.Required(
            CONF_TRIP_ADDITION,
            default=user_input.get(CONF_TRIP_ADDITION, 0),
        ): NumberSelector(
            NumberSelectorConfig(
                min=0, max=100, step=0.01, unit_of_measurement=UnitOfLength.KILOMETERS
            ),
        ),
        vol.Required(CONF_MULTIPLIERS): section(
            vol.Schema(
                {
                    vol.Required(
                        CONF_NEIGHBORHOOD,
                        default=user_input.get(CONF_MULTIPLIERS, {}).get(
                            CONF_NEIGHBORHOOD, 1
                        ),
                    ): NumberSelector(
                        NumberSelectorConfig(
                            min=0, step=0.01, mode=NumberSelectorMode.BOX
                        )
                    ),
                    vol.Required(
                        CONF_LOCAL,
                        default=user_input.get(CONF_MULTIPLIERS, {}).get(CONF_LOCAL, 1),
                    ): NumberSelector(
                        NumberSelectorConfig(
                            min=0, step=0.01, mode=NumberSelectorMode.BOX
                        )
                    ),
                    vol.Required(
                        CONF_HIGHWAY,
                        default=user_input.get(CONF_MULTIPLIERS, {}).get(
                            CONF_HIGHWAY, 1
                        ),
                    ): NumberSelector(
                        NumberSelectorConfig(
                            min=0, step=0.01, mode=NumberSelectorMode.BOX
                        )
                    ),
                }
            ),
            {"collapsed": True},
        ),
    }


class MovementConfigFlow(ConfigFlow, domain=DOMAIN):  # type: ignore[call-arg]
    """Handle a Movement config flow."""

    VERSION = 1

    def _user_form_schema(self) -> vol.Schema:
        return vol.Schema(_base_schema({}))

    @staticmethod
    @callback
    def async_get_options_flow(config_entry: ConfigEntry) -> OptionsFlow:
        """Get the options flow for this handler."""
        return MovementOptionsFlow()

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle a flow initialized by the user."""
        if user_input is not None:
            self._async_abort_entries_match(user_input)

            title = cast(
                State, self.hass.states.get(user_input[CONF_TRACKED_ENTITY])
            ).name

            slugified_existing_entry_titles = [
                slugify(e.title) for e in self._async_current_entries()
            ]

            possible_title = title
            tries = 1
            while slugify(possible_title) in slugified_existing_entry_titles:
                tries += 1
                possible_title = f"{title} {tries}"

            return self.async_create_entry(title=possible_title, data=user_input)

        return self.async_show_form(
            step_id="user",
            data_schema=self._user_form_schema(),
        )


class MovementOptionsFlow(OptionsFlow):
    """Handle a option flow."""

    def _user_form_schema(self, user_input: dict[str, Any]) -> vol.Schema:
        entity_registry = er.async_get(self.hass)
        return vol.Schema(
            {
                **_base_schema(user_input),
                vol.Required(SECTION_ADVANCED_OPTIONS): section(
                    vol.Schema(
                        {
                            vol.Required(
                                CONF_DEPENDENT_ENTITIES,
                                default=user_input.get(CONF_DEPENDENT_ENTITIES, []),
                            ): EntitySelector(
                                EntitySelectorConfig(
                                    integration=TEMPLATE_DOMAIN,
                                    domain=[SENSOR_DOMAIN],
                                    include_entities=[
                                        entity_id
                                        for entity_id in entity_registry.entities.keys()
                                        if _is_total_increasing_and_kilometers(
                                            entity_registry, entity_id
                                        )
                                    ],
                                    multiple=True,
                                ),
                            )
                        }
                    ),
                    {"collapsed": True},
                ),
            }
        )

    async def async_step_init(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle options flow."""
        if user_input is not None:
            advanced_options = user_input.pop(SECTION_ADVANCED_OPTIONS)
            self.hass.config_entries.async_update_entry(
                self.config_entry,
                data={**self.config_entry.data, **user_input, **advanced_options},
            )
            return self.async_create_entry(title=self.config_entry.title, data={})

        return self.async_show_form(
            step_id="init",
            data_schema=self._user_form_schema(dict(self.config_entry.data)),
        )


def _is_total_increasing_and_kilometers(
    entity_registry: er.EntityRegistry,
    entity_id: str,
) -> bool:
    """Determine if an entity_id represents an entity with a `state_class` of
    `total_increasing` and a `unit_of_measurement` of `km`."""
    return bool(
        (entry := entity_registry.async_get(entity_id))
        and (
            entry.capabilities
            and entry.capabilities.get("state_class")
            == SensorStateClass.TOTAL_INCREASING
            and entry.unit_of_measurement == UnitOfLength.KILOMETERS
        )
    )
