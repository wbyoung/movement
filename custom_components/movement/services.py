"""Support for the Movement integration."""

from functools import partial
from typing import Final

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.core import (
    HomeAssistant,
    ServiceCall,
    ServiceResponse,
    SupportsResponse,
    callback,
)
from homeassistant.exceptions import ServiceValidationError
from homeassistant.helpers import selector
import voluptuous as vol

from .const import DOMAIN
from .coordinator import MovementUpdateCoordinator
from .types import ModeOfTransit, MovementConfigEntryRuntimeData, ServiceAdjustment

SERVICE_NAME_ADD_DISTANCE: Final = "add_distance"
ATTR_ADJUSTMENTS: Final = "adjustments"
ATTR_ATTRIBUTES: Final = "attributes"
ATTR_CONFIG_ENTRY: Final = "config_entry"
ATTR_DISTANCE: Final = "distance"
ATTR_MODE_OF_TRANSIT: Final = "mode_of_transit"
ATTR_STATE: Final = "state"
ATTR_TRIGGER: Final = "trigger"


SERVICE_SCHEMA_ADD_DISTANCE: Final = vol.Schema(
    {
        vol.Required(ATTR_CONFIG_ENTRY): selector.ConfigEntrySelector(
            {
                "integration": DOMAIN,
            }
        ),
        vol.Optional(ATTR_DISTANCE): vol.Coerce(float),
        vol.Optional(ATTR_ADJUSTMENTS): vol.Coerce(float),
        vol.Optional(ATTR_MODE_OF_TRANSIT): vol.Coerce(ModeOfTransit),
    }
)


def _get_coordinator(
    hass: HomeAssistant, call: ServiceCall
) -> MovementUpdateCoordinator:
    """Get the coordinator from the entry."""

    entry_id: str = call.data[ATTR_CONFIG_ENTRY]
    entry: ConfigEntry | None = hass.config_entries.async_get_entry(entry_id)

    if not entry:
        raise ServiceValidationError(
            translation_domain=DOMAIN,
            translation_key="invalid_config_entry",
            translation_placeholders={
                "config_entry": entry_id,
            },
        )

    if entry.state != ConfigEntryState.LOADED:
        raise ServiceValidationError(
            translation_domain=DOMAIN,
            translation_key="unloaded_config_entry",
            translation_placeholders={
                "config_entry": entry.title,
            },
        )

    runtime_data: MovementConfigEntryRuntimeData = entry.runtime_data

    return runtime_data.coordinator


async def _add_distance(
    call: ServiceCall,
    *,
    hass: HomeAssistant,
) -> ServiceResponse:
    coordinator = _get_coordinator(hass, call)

    adjustment = ServiceAdjustment(
        distance=call.data.get(ATTR_DISTANCE, 0),
        adjustments=call.data.get(ATTR_ADJUSTMENTS, 0),
        mode_of_transit=call.data.get(ATTR_MODE_OF_TRANSIT),
    )

    await coordinator.async_perform_service_adjustment(adjustment)


@callback
def async_setup_services(hass: HomeAssistant) -> None:
    """Set up Movement services."""

    hass.services.async_register(
        DOMAIN,
        SERVICE_NAME_ADD_DISTANCE,
        partial(_add_distance, hass=hass),
        schema=SERVICE_SCHEMA_ADD_DISTANCE,
        supports_response=SupportsResponse.NONE,
    )
