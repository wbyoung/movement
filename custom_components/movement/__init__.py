"""The Movement integration."""

from __future__ import annotations

import logging

from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.event import (
    async_track_entity_registry_updated_event,
    async_track_state_change_event,
)
from homeassistant.helpers.typing import ConfigType

from .const import CONF_DEPENDENT_ENTITIES, CONF_TRACKED_ENTITY, DOMAIN
from .coordinator import MovementUpdateCoordinator
from .services import async_setup_services
from .types import MovementConfigEntry, MovementConfigEntryRuntimeData

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [Platform.SENSOR]
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)


async def async_setup(  # noqa: RUF029
    hass: HomeAssistant,
    config: ConfigType,  # noqa: ARG001
) -> bool:
    """Set up Movement services.

    Returns:
        If the setup was successful.
    """
    async_setup_services(hass)

    return True


async def async_setup_entry(hass: HomeAssistant, entry: MovementConfigEntry) -> bool:
    """Set up Movement from a config entry.

    Returns:
        If the setup was successful.
    """
    _LOGGER.debug("setup %s with config:%s", entry.title, entry.data)

    coordinator = MovementUpdateCoordinator(hass, entry)

    entry.async_on_unload(
        async_track_state_change_event(
            hass,
            [entry.data[CONF_TRACKED_ENTITY]],
            coordinator.async_handle_entity_state_change,
        ),
    )

    entry.async_on_unload(
        async_track_entity_registry_updated_event(
            hass,
            [
                entry.data[CONF_TRACKED_ENTITY],
                *entry.data.get(CONF_DEPENDENT_ENTITIES, []),
            ],
            coordinator.async_handle_tracked_entity_change,
        ),
    )

    await coordinator.async_config_entry_first_refresh()

    entry.runtime_data = MovementConfigEntryRuntimeData(
        coordinator=coordinator,
    )

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
    return True


async def async_unload_entry(hass: HomeAssistant, entry: MovementConfigEntry) -> bool:
    """Unload a config entry.

    Returns:
        If the unload was successful.
    """
    unload_ok: bool = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)

    if unload_ok:
        coordinator = entry.runtime_data.coordinator
        coordinator.cancel_all_listeners()

    return unload_ok


async def _async_update_listener(
    hass: HomeAssistant,
    entry: MovementConfigEntry,
) -> None:
    """Handle options update."""
    await hass.config_entries.async_reload(entry.entry_id)
