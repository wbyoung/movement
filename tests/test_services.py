"""Test services for the Movement integration."""

import datetime as dt
from enum import Flag
import re
from typing import Any, cast

from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ServiceValidationError
from homeassistant.util import dt as dt_util
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry
from syrupy.assertion import SnapshotAssertion
import voluptuous as vol

from custom_components.movement.const import DOMAIN
from custom_components.movement.services import (
    ATTR_ADJUSTMENTS,
    ATTR_CONFIG_ENTRY,
    ATTR_DISTANCE,
    ATTR_MODE_OF_TRANSIT,
    SERVICE_NAME_ADD_DISTANCE,
)
from custom_components.movement.types import (
    HistoryEntry,
    Location,
    ModeOfTransit,
    MovementData,
)

from . import MockNow


class DefaultConfigEntry(Flag):
    _singleton = True


USE_CONFIG = DefaultConfigEntry._singleton


@pytest.fixture
def config_entry_data(
    mock_config_entry: MockConfigEntry, request: pytest.FixtureRequest
) -> dict[str, str]:
    """Fixture for the config entry."""
    if "config_entry" in request.param and request.param["config_entry"] is USE_CONFIG:
        return {"config_entry": mock_config_entry.entry_id}

    return cast(dict[str, str], request.param)


@pytest.mark.usefixtures("init_integration")
async def test_has_services(
    hass: HomeAssistant,
) -> None:
    """Test the existence of the Movement Service."""
    assert hass.services.has_service(DOMAIN, SERVICE_NAME_ADD_DISTANCE)


@pytest.mark.usefixtures("init_integration")
@pytest.mark.parametrize(
    ("attrs",),
    [
        ({ATTR_DISTANCE: 1.2},),
        ({ATTR_ADJUSTMENTS: 3.1},),
        (
            {
                ATTR_DISTANCE: 1.2,
                ATTR_ADJUSTMENTS: 3.1,
                ATTR_MODE_OF_TRANSIT: str(ModeOfTransit.BIKING),
            },
        ),
    ],
    ids=["distance_only", "adjustments_only", "all_attrs"],
)
async def test_add_distance_service(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_now: MockNow,
    snapshot: SnapshotAssertion,
    attrs: dict[str, Any],
):
    """Test `add_distance` service call."""
    coordinator = mock_config_entry.runtime_data.coordinator
    coordinator.inject_data(
        data=MovementData(
            distance=0,
            adjustments=0,
            speed=0,
            mode_of_transit=ModeOfTransit.WALKING,
            change_count=0,
            ignore_count=0,
        ),
        history=[
            HistoryEntry(
                at=dt_util.utcnow() - dt.timedelta(seconds=51.23),
                location=Location(55.301973, -114.820543),
                accuracy=44,
            )
        ],
        transition=None,
    )

    await hass.services.async_call(
        DOMAIN,
        SERVICE_NAME_ADD_DISTANCE,
        {
            ATTR_CONFIG_ENTRY: mock_config_entry.entry_id,
            **attrs,
        },
        blocking=True,
    )

    assert hass.states.get("sensor.mock_title_distance_traveled") == snapshot


@pytest.mark.usefixtures("init_integration")
@pytest.mark.parametrize(
    ("config_entry_data", "service_data", "error", "error_message"),
    [
        ({}, {}, vol.er.Error, "required key not provided .+"),
        (
            {"config_entry": USE_CONFIG},
            {"distance": "incorrect distance value"},
            vol.er.Error,
            "expected float for dictionary value .+",
        ),
        (
            {"config_entry": USE_CONFIG},
            {"adjustments": "incorrect adjustments value"},
            vol.er.Error,
            "expected float for dictionary value .+",
        ),
        (
            {"config_entry": USE_CONFIG},
            {"mode_of_transit": "incorrect mode_of_transit value"},
            vol.er.Error,
            "expected ModeOfTransit or one of 'walking', 'biking', 'driving' for dictionary value .+",
        ),
        (
            {"config_entry": "incorrect entry"},
            {},
            ServiceValidationError,
            "Invalid config entry.+",
        ),
    ],
    indirect=["config_entry_data"],
)
async def test_add_distance_service_validation(
    hass: HomeAssistant,
    config_entry_data: dict[str, str],
    service_data: dict[str, str],
    error: type[Exception],
    error_message: str,
) -> None:
    """Test the `add_distance` service validation."""

    with pytest.raises(error) as exc:
        await hass.services.async_call(
            DOMAIN,
            SERVICE_NAME_ADD_DISTANCE,
            config_entry_data | service_data,
            blocking=True,
        )
    assert re.match(error_message, str(exc.value))


@pytest.mark.usefixtures("init_integration")
async def test_add_distance_service_called_with_unloaded_entry(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test `add_distance` service call with unloaded config entry."""
    await hass.config_entries.async_unload(mock_config_entry.entry_id)

    with pytest.raises(ServiceValidationError):
        await hass.services.async_call(
            DOMAIN,
            SERVICE_NAME_ADD_DISTANCE,
            {"config_entry": mock_config_entry.entry_id},
            blocking=True,
        )
