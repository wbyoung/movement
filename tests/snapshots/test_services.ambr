# serializer version: 1
# name: test_add_distance_service[adjustments_only]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'adjustments': 3.1,
      'device_class': 'distance',
      'friendly_name': 'Mock Title Distance traveled',
      'icon': 'mdi:map-marker-distance',
      'ignore_count': 0,
      'mode_of_transit': None,
      'speed': 0,
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
      'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
      'update_rate': None,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_distance_traveled',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '3.1',
  })
# ---
# name: test_add_distance_service[all_attrs]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'adjustments': 3.1,
      'device_class': 'distance',
      'friendly_name': 'Mock Title Distance traveled',
      'icon': 'mdi:map-marker-distance',
      'ignore_count': 0,
      'mode_of_transit': <ModeOfTransit.BIKING: 'biking'>,
      'speed': 0,
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
      'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
      'update_rate': None,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_distance_traveled',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '4.3',
  })
# ---
# name: test_add_distance_service[distance_only]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'adjustments': 0,
      'device_class': 'distance',
      'friendly_name': 'Mock Title Distance traveled',
      'icon': 'mdi:map-marker-distance',
      'ignore_count': 0,
      'mode_of_transit': None,
      'speed': 0,
      'state_class': <SensorStateClass.TOTAL_INCREASING: 'total_increasing'>,
      'unit_of_measurement': <UnitOfLength.KILOMETERS: 'km'>,
      'update_rate': None,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.mock_title_distance_traveled',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1.2',
  })
# ---
